### ISSUE TITLE
CLI & Eyre Freezing for no apparent reason

### ISSUE THREAD

**Issue by liam-fitzgerald** [urbit/urbit#1981]
Every now and again ~hastuc-dibtux freezes totally. The CLI will lockup and ignore any input. Attempting to access landscape times out. No apparent error logs or anything like that. Ctrl-Z and relaunching is the only apparent fix.

**Comment by jtobin** [comment]
@liam-fitzgerald Has this resurfaced at all in the new era?

**Comment by philipcmonk** [comment]
Yes I've seen this several times this era. I've even seen it during event
replay.

**Comment by joemfb** [comment]
`Ctrl=\` or `kill -3` will produce a coredump in this case. Or you could attach a debugger. (Wait for the next hotfix when the release binaries will have debug symbols.) I expect both processes are just hanging in the event loop, but it'd be good to confirm.

**Comment by joemfb** [comment]
I strongly suspect that this is fixed, at least partially, in #2042. That fixes a case in IPC message parsing where the buffer read length wasn't properly recorded. It would continue waiting for bytes it had already read, which would cause IPC to appear to hang.

**Comment by liam-fitzgerald** [comment]
FWIW, I haven't seen it this era

**Comment by litpub** [comment]
I _think_ I am seeing this on ~nopweb-nopbud. 

Symptoms:
- CLI freezes totally, no input is recognized
- CLI reports no activity, no behn/ames etc
- Landscape timesout
- Channels do not work

When I forced ctrl+z I get:

```
Assertion failed: !(handle->flags & (UV_CLOSING | UV_CLOSED)) (src/unix/core.c: uv_close: 103)
address 0 out of loom!
loom: [0x200000000 : 0x280000000)
Assertion '0' failed in noun/events.c:129

bail: oops
bailing out
Assertion failed: 0 (noun/manage.c: u3m_bail: 667)
Segmentation fault (core dumped)
```

I'm running the 0.10.3 binary.

**Comment by ryjm** [comment]
this happened on ~littel-wolfur as well over the last couple days, but seemed to fix itself after a couple of restarts.

**Comment by galenwp** [comment]
Can confirm, this happens to me regularly. It actually seems like it is happening *more and more* frequently.

**Comment by matildepark** [comment]
I had it a lot when OS1 first OTA’d! Not for a week now. For a while I just got used to ssh-ing in and ctrl+z. (I know — I should’ve reported.)

—
~haddef-sigwen
https://urbit.org

**Comment by galenwp** [comment]
I had a feeling this was probably somewhat widespread...

**Comment by loganallenc** [comment]
This happened to me last night. I’ve been trying for the life of me to figure out why

**Comment by loganallenc** [comment]
I'm having issues getting my ship to respond to keyboard input at all, even after starting and stopping it multiple times *while it isn't even spinning / behning / etc*

**Comment by belisarius222** [comment]
Does this ever resolve, or does it hang forever?  If it does resolve, maybe the ship is taking a snapshot, which I could imagine being slow if it's right after a |pack defragmentation, but I can't remember whether that gets run automatically.

—
~rovnys-ricfer
https://urbit.org

**Comment by maxwell-black** [comment]
I too have been having this problem. I did a personal breach and that seemed to work, but certain apps like the chat client are still not working.

**Comment by liam-fitzgerald** [comment]
I’ve been reproducing this again, although eyre appears to work when the CLI is locked up now. Unsure if this is caching hiding things or something else. 

FWIW @belisarius222 , I just restarted my ship again and it had spent the better part of today locked up (8 hours roughly)

**Comment by galenwp** [comment]
Mine seems to almost certainly hang forever. It'll sit like this for hours before I have to kill and restart it. It'll always print:

```
Assertion '!"exit-tcsetattr"' failed in vere/term.c:280

bail: oops
bailing out
Assertion failed: 0 (noun/manage.c: u3m_bail: 667)
```

On restart.

I thought maybe it was because I was still on 10.3 — but it still happens on 10.4.

**Comment by galenwp** [comment]
I appear to be in a loop of experiencing this issue. It seems to be somehow triggered by the most recent OTA. The ship will spin for a while on ` behn`, give me a dojo for a little bit, then hang. After being killed, it'll restart and print the `tcsetattr` error. 

This did *seem* to get somewhat better with 10.4, but it's still regular enough that it drives me crazy.

**Comment by joemfb** [comment]
I'm not sure that these are all the same issue. @litpub's error is a double-close of a libuv handle, while @galenwp's is a failed terminal-configuration call (most likely when we hijack the terminal for printf logging).

For anyone still experiencing this, can you confirm that you're running urbit v0.10.4, and whether the hang/crash output looks like @litpub's, @galenwp's, or something else?

**Comment by mocknotes** [comment]
> Every now and again ~hastuc-dibtux freezes totally. The CLI will lockup and ignore any input. Attempting to access landscape times out. No apparent error logs or anything like that. Ctrl-Z and relaunching is the only apparent fix.

This is happening for me. Think it started after the most recent OTA. urbit-os-v1.0.13: 0v1e.va6cc.cks0i.8dhed.sre6p.pi67c.4b9he.d9c5s.ii3v2.c2v76.i2ffu

**Desktop (please complete the following information):**
 - OS: MacOS 10.15.4
 - Browser: Chrome / Terminal
 - 0v1e.va6cc.cks0i.8dhed.sre6p.pi67c.4b9he.d9c5s.ii3v2.c2v76.i2ffu

**Additional context**
Ctrl + D does not respond to shut down Urbit.
Ctrl + Z to exit and relaunch Urbit fixes it until it happens again spontaneously.

**Comment by joemfb** [comment]
Thanks, @mocknotes. Is there any consistent pattern of output before this happens?

**Comment by mocknotes** [comment]
@joemfb None that I'm aware of. 

I don't have logs for previous times it occurred since I closed my terminal since then. The last message before freeze is the following (which I do see from time to time without freezing):

`[%mailbox-kick i=%~.mailbox t=[i=~.~ t=/~sanpel-palnet/example-room-name]]
%chat-hook-resubscribe`

Ctrl + Z to exit gave me:

`[received keyboard stop signal, exiting]`

Restarting urbit functions as expected. I do not receive the messages galen or litpub reported.

**Comment by liam-fitzgerald** [comment]
This has been popping up for me again, but with a twist. Eyre works fine, but the CLI does not accept any input. Ctrl+D does nothing, and Ctrl+Z makes the quit process hang indefinitely until i kill -9 the urbit process

**Comment by joemfb** [comment]
@liam-fitzgerald, can you try running with the changes in #2919? It's compatible with the livenet.

**Comment by liam-fitzgerald** [comment]
looks good so far, will let you know if I run into anything further

**Comment by lukechampine** [comment]
I'm experiencing similar problems. My ship is currently frozen with prompt:
```
~watter-parter:dojo> +
```
Does not accept any input, except Ctrl-Z. I can also Ctrl-C sometimes, e.g. when «behn»ing. Restarting brings me back to the frozen `+` prompt. Output works fine. Landscape works fine, including dojo -- except for Chat, which gives a javascript error:
```
window.urb.setOnChannelError is not a function. (In 'window.urb.setOnChannelError(this.onChannelError.bind(this))', 'window.urb.setOnChannelError' is undefined)
```
In fact, I found that `%chat-cli` was not running. I was able to start it from web dojo, but not `%chat-store` or `%chat-view`; they fail with `-find.^configs` and `-find.^action`. Strangely, the unix mount does not appear to work: if I create or change a file and run `|commit %home` (from web dojo), the change is not reflected in Clay. Sadly, this means I can't attempt to fix things by manually editing .hoon files.

Tried running the silver bullet via web dojo, no change. Attempted to replay from 0 (with both the standard binary and @joemfb's special replay binary), and got started failing with `bail: meme` (IIRC) around event 160000 (out of ~8mil total).
 
This is on base hash `r92oj`, release v0.10.4, on linux. Also tried running `release/next-vere`, no change.

It's looking like a breach is the only way to fix this, unless I can successfully replay from scratch. Although tbh, I suspect that even a full replay would just bring me right back to a frozen prompt... 😓

**Comment by alecbrick** [comment]
I started experiencing these issues last night. My terminal started `behn`ing as if I was getting an OTA. This is odd, since the latest base hash (`7eqpf`) hasn't changed. Afterwards, I ended up in the same state as other people in this thread. I can't type anything into the terminal, and Landscape just gives a mostly blank screen (except for the Groups page, which I can click but not see any groups).

I'm on binary 0.10.7, with OS version 1.0.23. Any help would be appreciated!
